lgc/util/CacheHashNoteUtils.cpp

```diff
@@ -0,0 +1,298 @@
+//                                new offset to write those sections.
+void getNewOffsetsForSections(const SmallVectorImpl<char> &elf, const ArrayRef<ELF::Elf64_Shdr> sectionHeaders,
+                              const ELF::Elf64_Off endOfSectionHeaderTable, ELF::Elf64_Off initialOffset,
+                              SmallVectorImpl<SectionShiftInfo> &sectionAndNewOffset) {
+  // If a section is located after the section header table (i.e., its offset
+  // is bigger than endOfSectionHeaderTable), it will be shifted. We keep the
+  // section's contents and its new offset. This information will be used when
+  // we rewrite the ELF.
+  for (auto it = sectionHeaders.begin(); it != sectionHeaders.end(); ++it) {
+    if (it->sh_offset < endOfSectionHeaderTable)
+      continue;
+    sectionAndNewOffset.push_back({SmallString<64>(StringRef(elf.data() + it->sh_offset, it->sh_size)),
+                                   alignTo(initialOffset, Align(it->sh_addralign))});
+    initialOffset = sectionAndNewOffset.back().newOffset + it->sh_size;
+  }
+}
+
+// =====================================================================================================================
+// Rewrite ELF to add the new note section and its header for the cache hash.
+//
+// @param [in/out] elf      : ELF to be rewritten with the new .note section
+//                          for the cache hash.
+// @param offsetOfNewSectionHeader   : The offset where the new section header will be
+//                          placed.
+// @param newNoteSectionHeader      : The section header for the new note section.
+// @param zerosForAlignment : Zeros for the note section alignment between the
+//                          end of section header table and the section for
+//                          cache hash note.
+// @param noteSectionStream       : The note secion containing the cache hash.
+// @param sectionAndNewOffset  : The information to shift sections after the
+//                          section header table.
+void rewriteELFWithCacheHash(SmallVectorImpl<char> &elf, const ELF::Elf64_Off offsetOfNewSectionHeader,
+                             const ELF::Elf64_Shdr &newNoteSectionHeader, const SmallString<4> &zerosForAlignment,
+                             AppendingBinaryByteStream &noteSectionStream,
+                             SmallVectorImpl<SectionShiftInfo> &sectionAndNewOffset) {
+  // Strip sections after the section header table.
+  elf.resize(offsetOfNewSectionHeader);
+
+  // Increase the number of section headers in ELF header.
+  reinterpret_cast<ELF::Elf64_Ehdr *>(elf.data())->e_shnum++;
+
+  // Write the section header for the new note section.
+  raw_svector_ostream elfStream(elf);
+  elfStream << StringRef(reinterpret_cast<const char *>(&newNoteSectionHeader), sizeof(ELF::Elf64_Shdr));
+
+  // Write zeros for alignment before the new section.
+  elfStream << zerosForAlignment.str();
```

Comment:
It is awkward to have to pass in a parameter that just for alignment.  Since the alignment requirements are part of elf, this function should be able to figure out what is needed for padding.

Can you se the "tell()" function in the `elfStream` or check the size of `elf` to determine the amount of padding that is needed?

Reply:
done